Sources/DataStructures/SegmentTree.swift

```diff
@@ -0,0 +1,130 @@
+//
+//  SegmentTree.swift
+//  DataStructures
+//
+//  Created by James Bean on 12/4/17.
+//
+
+import Combinatorics
+import Destructure
+
+public struct Interval <Metric: SignedNumeric & Comparable> {
+
+    enum Bound {
+        case start
+        case end
+    }
+
+    let start: Metric
+    let end: Metric
+
+    public init(_ start: Metric, _ end: Metric) {
+        precondition(end >= start)
+        self.start = start
+        self.end = end
+    }
+}
+
+extension Interval: Equatable {
+    public static func == (lhs: Interval, rhs: Interval) -> Bool {
+        return lhs.start == rhs.start && lhs.end == rhs.end
+    }
+}
+
+enum SegmentTree <Metric: SignedNumeric & Comparable> {
```

Comment:
where does `SegmentTree` fit into this design? i think this is a pretty interesting representation. wondering if duplicating the LHS and RHS bounds is necessary at every level; is that a simplicity thing? because you could also define interval recursively as (left subtree's left bound, right subtree's right bound) in the branch case